### ISSUE TITLE
useSharedValue reinitialize on re-render

### ISSUE THREAD

**Issue by ethanshar** [software-mansion/react-native-reanimated#3224]
## Description

It appears `useSharedValue` reset the value on each render of a component, instead of setting it once with the initial value. 
I notice this issue https://github.com/software-mansion/react-native-reanimated/issues/1655 but wasn't 100% sure if this is something that was fixed. 

### Expected behavior
`useSharedValue` should initialize the value once with provided initial value. 

### Actual behavior & steps to reproduce
Run the snippet below to reproduce

## Snack or minimal code example

```
function getInitialValue() {
  console.log('this is being called on each render')
  return [];
}

function MyComponent() {
  const myData = useSharedValue(getInitialValue())  

  useAnimatedReaction(() => myData.value,
    () => {
      console.log('myData has changed');
    });

  return null;
}

function ParentComponent() {
  const [counter, setCounter] = useState();

  return <View>
    <TouchableOpacity onPress={() => setCounter(counter + 1)}>
       <Text>{counter}</Text>
    </TouchableOpacity>
    <MyComponent/>
  </View>
}
```

## Package versions

| name                                 | version |
| ------------------------------------ | ------- |
| react-native                         |  0.66.4       |
| react-native-reanimated              |  2.8.0        |
| NodeJS                               |     14.17.0    |
| Xcode                                |     13.1    |
| Java                                 |         |
| Gradle                               |         |
| expo                                 |         |

## Affected platforms

- [x] Android
- [x] iOS
- [ ] Web

**Comment by nandorojo** [comment]
yeah, as @andreialecu, by calling the function, you are calling it on each render. the behavior is correct.

`useState` lets you pass a function and under the hood it only calls it for you once. `useSharedValue` doesn't have a lazy API, I believe. Maybe you can do something like this:

```ts
const value = useSharedValue(null)

if (value.value === null) {
  value.value = getValueLazily()
}
```

I'm not sure if this is safe for concurrent mode or not. It may be fine.

**Comment by ethanshar** [comment]
I would expect `useSharedValue` to work similarly to `useState` or `useRef`.
Creating a variable using one of these hooks preserve its value through renders. 

Having to memoize the shared value sound trivial to me, why would anyone **don't want** to memoize it? 
Shouldn't it be memoized by the hook internally? 
Does it make sense for a shared value to reset its value on each render? to me It sounds like a bug.

**Comment by nandorojo** [comment]
`useAnimationReaction` receives a `new` and `old` argument. you can compare these two in the function and only do something if it has changed.

**Comment by ethanshar** [comment]
> `useAnimationReaction`‘s function receives both a `new` and `old` argument. you can compare these two in the function and only do something if it has changed.

Yes, I'm well aware and that's exactly the problem I'm facing. 
Because the shared value is being reset on render, I receive a reaction on this change

**Comment by andreialecu** [comment]
@ethanshar re-read my message just above, it explains why this happens and has nothing to do with the shared value. There's also a snack you can experiment it.

`useAnimatedReaction(() => 0, () => console.log('myData has changed'));` prints the same message

**Comment by dalmendray** [comment]
@ethanshar did you find a work around for this? I'm facing the same issue with useSharedValue and useAnimatedStyle

**Comment by vkukade-altir** [comment]
Unfortunately, this issue still exists. Shared value gets reset on every re-render of component!

**Comment by zay1x** [comment]
**Explain:** Each time component is re-render. **sharedValue** has not been updated yet. It reset to init value. So, you can update it after your component done for re-render.

```
interface Props {
   counter: number;
    ...
};

const Child = () => {
    const sharedValue = useSharedValue(0);

    useEffect(() => {
        const timeoutId = setTimeOut(() => {    // <= Here is solution for me
            sharedValue.value = counter;
        }, 0);
    }, [counter]);

    return <View>
        <Text>{counter}</Text>
    </View>
};

export default memo(Child);
```

**Comment by WayneKim92** [comment]
I got the same issue.

**Comment by tjzel** [comment]
Some time ago `useSharedValue` started to use `useState` under the hood which doesn't reinitialize value unless the component has been remounted. I'm closing this issue since I don't see any reproductions here that apply to the current implementation of `useSharedValue`. If the issue persists for you on the current version of Reanimated and you have a reproduction, feel free to open a new issue.
